src/porepy/viz/exporter.py

```diff
@@ -520,546 +1178,889 @@ def _export_grid(
 
-            pts_pos += g.num_nodes
+            # Update offsets
+            nodes_offset += g.num_nodes
+            cell_offset += g.num_cells
 
-        # construct the meshio data structure
-        num_block = len(cell_to_nodes)
-        meshio_cells = np.empty(num_block, dtype=object)
-        meshio_cell_id = np.empty(num_block, dtype=object)
+        # Construct the meshio data structure
+        num_blocks = len(cell_to_nodes)
+        meshio_cells = np.empty(num_blocks, dtype=object)
+        meshio_cell_id = np.empty(num_blocks, dtype=object)
 
+        # For each cell_type store the connectivity pattern cell_to_nodes for
+        # the corresponding cells with ids from cell_id.
         for block, (cell_type, cell_block) in enumerate(cell_to_nodes.items()):
-            # remove the number of nodes associated to the polygon since meshio requires only
-            # the keyword "polygon" for polygons
-            cell_type_ = "polygon" if "polygon" in cell_type else cell_type
-            meshio_cells[block] = meshio.CellBlock(cell_type_, cell_block.astype(int))
+
+            # Meshio requires the keyword "polygon" for general polygons.
+            # Thus, remove the number of nodes associated to polygons.
+            cell_type_meshio_format = "polygon" if "polygon" in cell_type else cell_type
+            meshio_cells[block] = meshio.CellBlock(
+                cell_type_meshio_format, cell_block.astype(int)
+            )
             meshio_cell_id[block] = np.array(cell_id[cell_type])
 
-        return meshio_pts, meshio_cells, meshio_cell_id
+        # Return final meshio data: points, cell (connectivity), cell ids
+        return Meshio_Geom(meshio_pts, meshio_cells, meshio_cell_id)
+
+    def _sort_point_pairs_numba(self, lines: np.ndarray) -> np.ndarray:
+        """
+        This a simplified copy of pp.utils.sort_points.sort_point_pairs.
+
+        Essentially the same functionality as sort_point_pairs, but stripped down
+        to the special case of circular chains. Differently to sort_point_pairs, this
+        variant sorts an arbitrary amount of independent point pairs. The chains
+        are restricted by the assumption that each contains equally many line segments.
+        Finally, this routine uses numba.
 
-    def _export_3d(
-        self, gs: Iterable[pp.Grid]
-    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        Parameters:
+            lines (np.ndarray): Array of size 2 * num_chains x num_lines_per_chain,
+                containing node indices. For each pair of two rows, each column
+                represents a line segment connectng the two nodes in the two entries
+                of this column.
+
+        Returns:
+            np.ndarray: Sorted version of lines, where for each chain, the collection
+                of line segments has been potentially flipped and sorted.
+        """
+
+        import numba
+
+        @numba.jit(nopython=True)
+        def _function_to_compile(lines):
```

Comment:
It should be possible to specify the data types of in- and output. That should speed up numba significantly. @jwboth, let's discuss this.

Reply:
Included now.